### ISSUE TITLE
FR: Add an option for generating string enum.

### ISSUE THREAD

**Issue by Rin0107** [bufbuild/protobuf-es#1032]
# TL;DR
I propose an option that generate string enum (instead of numeric enum).

For example, the current enum generation looks like this:
```typescript
enum Role {
  UNSPECIFIED = 0,
  ADMIN = 1,
  INTERNAL_ADMIN = 2,
  CONSUMER = 3
}
```
With the proposed string enum:
```typescript
enum Role {
  UNSPECIFIED = ‘UNSPECIFIED’,
  INTERNAL_ADMIN = ‘INTERNAL_ADMIN’
  CONSUMER = ‘CONSUMER’
}
```

## Why
Numeric enum in TS has a problem. It is type unsafe.
It is allowed below access.
```typescript
enum Role {
  UNSPECIFIED = 0,
  ADMIN = 1,
  INTERNAL_ADMIN = 2,
  CONSUMER = 3
}

console.log(Role[4]) // It is expected compile error but…
=> undefiend
```
String enums can check it and occur compile error.

Then, in below use-case, only-numeric-enums-generate cause a problem.
I face it in a situation like that I generated ts code in BFF (Backend for Frontend) from Backend protobuf.
ClientSide and Backend expect string (e.g ‘ADMIN’), but BFF has only numeric enum, so BFF has to re-defien string enum by myself.
```typescript
enum RoleByMyself {
  UNSPECIFIED = ‘UNSPECIFIED’,
  INTERNAL_ADMIN = ‘INTERNAL_ADMIN’
  CONSUMER = ‘CONSUMER’
}

console.log(RoleByMyself[‘INTERNAL_ADMIN’])
=> ‘INTERNAL_ADMIN’
console.log(RoleByMyself[‘foo’]) // As expected, compile error 
```

For reference, [stephenh/ts-proto](https://github.com/stephenh/ts-proto) has this option [stringEnums=true](https://github.com/stephenh/ts-proto#:~:text=With%20%2D%2Dts_proto_opt%3DstringEnums%3Dtrue%2C%20the%20generated%20enum%20types%20will%20be%20string%2Dbased%20instead%20of%20int%2Dbased.).

Thanks for reading!

**Comment by timostamm** [comment]
Hey Rin, 

regarding type-safety: TypeScript enums provide an index signature that behaves like any other index signature. For example:

```ts
const arr: string[] = [];
const val: string = arr[1]; // actually undefined
```

You can configure the behavior with the compiler option [noUncheckedIndexedAccess](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess). TypeScript are certainly not perfect, but I'm not sure this specific issue is the most problematic, or that it warrants switching to string enums - which don't even provide a mapping from numeric values at all.

Regarding a BFF, can you provide more details on the situation? What format exactly are BFF and F using? I suspect that there's already a solution.

**Comment by Rin0107** [comment]
Thanks for replying!

Ok, I got your description of behaves about index signature. 
For notes, [this post](https://dev.to/ivanzm123/dont-use-enums-in-typescript-they-are-very-dangerous-57bh) speaks for my opinion.

And I sincerely appreciate your kindness in asking my situation.
Tomorrow, I will provide a more detailed explanation of it in my next comment.
